### ISSUE TITLE
Tbb blocked range size of 1?

### ISSUE THREAD

**Issue by mel255** [tiny-dnn/tiny-dnn#389]
Hi,

I'm unable to get tiny-dnn to use multiple threads with TBB. When I use a profiler on the provided benchmark (tiny-dnn/examples/benchmarks/main.cpp), the maximum number of threads ever active at one time is always one, even when I initialize a task scheduler inside parallel_for.h with a set number of threads much larger than one.

To look into this problem, I added a print statement within the CNN_USE_TBB parallel_for (inside parallel_for.h) to print the value of the arguments (begin, end, and grain size). "End" always prints out as 1, which means that the call to tbb::parallel_for always has a blocked range size of 1. Why is the size of the blocked range always 1? How could a blocked range of size 1 be divided up into multiple tasks that could fully exploit TBB? I think this is related to my problem over only observing one active thread.

Thanks!

**Comment by edgarriba** [comment]
@mel255 have you compiled with `-DUSE_TBB=ON` ?
I have monitored my cpus and all are activated with tbb.

**Comment by mel255** [comment]
Yes, I have compiled with -DUSE_TBB=ON and I verify that the cmake summary indicates TBB is found and in use. How are you verifying that all your cores are being used, and do you have a sample network that you are testing this on?

**Comment by pliptor** [comment]
Hi, 

On Linux I can tell TBB is indirectly on by looking at the CPU% field with the `top` utility. It usually runs at around 300 to 400% (quad core CPU) in my case when TBB is on and a maximum of 100% with TBB off.

I also applied a patch in my build as you describe to replicate your tests.

~~~
diff --git a/tiny_dnn/util/parallel_for.h b/tiny_dnn/util/parallel_for.h
index ec10bb0..92d21e1 100644
--- a/tiny_dnn/util/parallel_for.h
+++ b/tiny_dnn/util/parallel_for.h
@@ -46,6 +46,7 @@ typedef tbb::blocked_range<int> blocked_range;
 
 template <typename Func>


 void parallel_for(int begin, int end, const Func &f, int grainsize) {
+  std::printf("test %d %d %d\n", begin, end, grainsize);
   tbb::parallel_for(
     blocked_range(begin, end, end - begin > grainsize ? grainsize : 1), f);
 }
~~~

I confirm that running tiny-dnn/examples/benchmarks/main.cpp  always leave `end=1` but I suspect the example might be too short for loop unrolling (not sure).

[benchmark.txt](https://github.com/tiny-dnn/tiny-dnn/files/843439/benchmark.txt)

Try running example_mnist_train ../data

[mnist.txt](https://github.com/tiny-dnn/tiny-dnn/files/843445/mnist.txt)

It does show `end` with values other than 1. Also, the training process finishes four times faster in my setup with TBB on. I hope this helps.
